src/bp/invoice.rs

```diff
@@ -0,0 +1,128 @@
+// LNP/BP Core Library implementing LNPBP specifications & standards
+// Written in 2020 by
+//     Dr. Maxim Orlovsky <orlovsky@pandoracore.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+use bitcoin::hashes::sha256d;
+use bitcoin::secp256k1::Signature;
+use bitcoin::Address;
+use miniscript::{descriptor::DescriptorPublicKey, Descriptor, Miniscript};
+use url::Url;
+
+use crate::bp::blind::OutpointHash;
+use crate::bp::chain::AssetId;
+use crate::bp::{Chain, HashLock, P2pNetworkId, Psbt, ScriptPubkeyFormat};
+use crate::lnp::{tlv, Features};
+use crate::secp256k1;
+
+#[derive(Tlv)]
+#[lnpbp_crate(crate)]
+pub enum FieldType {
+    #[tlv(type = 0x01)]
+    Payers,
+}
+
+#[derive(Lnp)]
+#[tlv_types(FieldType)]
+#[lnpbp_crate(crate)]
+pub struct Invoice {
+    network: P2pNetworkId,
+
+    /// List of beneficiary dests ordered in most desirable first order
+    beneficiaries: Vec<Beneficiary>,
+
+    /// Optional list of payers authored to pay
+    #[tlv(type = FieldType::Payers)]
+    payers: Vec<Payer>,
+    quantity: Quantity,
+    price: Option<AmountExt>,
+
+    /// If the price of the asset provided by fiat provider URL goes below this
+    /// limit the merchant will not accept the payment and it will become
+    /// expired
+    fiat_requirement: Option<Fiat>,
```

Comment:
Probably "expired" is not correct term (we have expiry date anyway, which is unrelated to this). So basically this indicates payer that if the current price is below this threshold the payment may be rejected by the merchant

Reply:
Today this situation is handled by merchants for on-chain txes anyway. This will just give them a standard way to communicate when they will be using their existing failback scenarios (like asking for the second payment on the difference, or returning payment - some fee).

I.e. today merchants have to write this on their website or on ATM, but with this this information can be also presented by the wallet like "the price is very volatile right now and there is a risk that merchant will not accept this payment"

Reply:
I got the chance to read Rusty's [bolt12 "Offer" proposal](https://github.com/lightningnetwork/lightning-rfc/pull/). It includes a non-lightning currency, but it's not in the invoice itself: it's in so-called offers, and the invoice itself only contains a msat value. This way a client should still able to know deterministically whether a given payment will be accepted as paying the invoice.
If a merchant sets a value for "fiat_requirement", then the client is not able to know beforehand if a given payment will be able to pay for the invoice, since exchange rate can change while the payment goes through.

So I'm not sure I understand this correctly, but this seems to me very different from Rusty's proposal.

Reply:
Yes, the approach is very different from the Rusty's, but it was his idea to take fiat into the account :)

Reply:
I'm still not sure it's a good idea to have an invoice design that does not allow payer to deterministically predict whether a payment will be accepted for a given invoice. This could lead to very bad experience if user pays onchain and needs to be refunded onchain, with possibly high fees and lock of funds for a significant amount of time.

I understand that this can happen anyways, but having it as a design choice will probably make this situation more common with (in my opinion) little or no advantages.

I don't want to take this discussion further, I just wanted to give my opinion for what it matters.

Reply:
I not sure I get your point, since this is exactly what we have here (and is absent in LN Offers): a payee can predict whether the merchant will accept the payment. We have an indicative exchange link to get the rate acceptable for merchant, and have a threshold rate after which the merchant MAY have to reject the payment.

Reply:
My point is that the rate will change while the payment goes through, and the payment might take hours to be confirmed. User and merchant will not look at the exchange rate in the same moment and this breaks determinism.

But again, it's my humble opinion. If you don't agree please ignore it.

Reply:
My take on this: a merchant must use software (like BtcPay Server) that should check the volatility index and adjust the threshold accordingly. Then merchant can guarantee the payment and merchant risk will be measurable & insurable (or transferrable to the client). Without the proposed mechanism this will be impossible (for the merchant to guarantee QoS for a long-standing invoices) and I see no other way of achieving this (correct me if I'm wrong)